mlx/array.cpp

```diff
@@ -192,6 +192,36 @@ array::ArrayDesc::ArrayDesc(
   init();
 }
 
+array::ArrayDesc::~ArrayDesc() {
+  // When an array description is destroyed it will delete a bunch of arrays
+  // that may also destory their corresponding descriptions and so on and so
+  // forth.
+  //
+  // This calls recursively the destructor and can result in stack overflow, we
+  // instead put them in a vector and destroy them one at a time resulting in a
+  // max stack depth of 2.
+  std::vector<std::shared_ptr<ArrayDesc>> for_deletion;
+
+  for (array& a : inputs) {
+    if (a.array_desc_ != nullptr && a.array_desc_.use_count() == 1) {
```

Comment:
If you want you can rely on the fact that `use_count == 0` if it's a `nullptr`, but maybe a little too slippery

```suggestion
    if (a.array_desc_.use_count() == 1) {
```

Reply:
Done. Very good point.